File: 605proj/benchmark.cpp
//Results will be saved to Cpp_CPU.txt and Cpp_GPU.txt
#include <iostream>
#include <fstream>
#include <Eigen/Dense>
#include <chrono>
#include <vector>
#include <cmath>
#include <string>
#include <sys/resource.h>
#include <numeric>

using namespace Eigen;
using namespace std;
class TransformerBlock {
public:
    TransformerBlock(int embed_dim, int ff_hidden_dim) {
        Wq   = MatrixXd::Random(embed_dim, embed_dim);
        Wk   = MatrixXd::Random(embed_dim, embed_dim);
        Wv   = MatrixXd::Random(embed_dim, embed_dim);
        Wff1 = MatrixXd::Random(embed_dim, ff_hidden_dim);
        Wff2 = MatrixXd::Random(ff_hidden_dim, embed_dim);
    }
    MatrixXd forward(const MatrixXd& x) {
        MatrixXd Q = x * Wq;
        MatrixXd K = x * Wk;
        MatrixXd V = x * Wv;
        MatrixXd scores = (Q * K.transpose()) / sqrt(double(Q.cols()));
        scores = scores.array().exp();
        scores = scores.array().colwise() / scores.rowwise().sum().array();
        MatrixXd context = scores * V;
        MatrixXd ff = (context * Wff1).unaryExpr([](double v){ return v > 0 ? v : 0; });
        ff = ff * Wff2;
        return ff;
    }
private:
    MatrixXd Wq, Wk, Wv, Wff1, Wff2;
};
static long getCurrentRSS_kB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss;
}
struct Config {
    string name;
    int embed_dim;
    int ff_hidden_dim;
};
int main() {
    const int batch_size = 32;
    const int seq_length = 128;
    const int runs       = 100;

    vector<Config> configs = {
            {"Small",  128,  512},
            {"Medium", 256, 1024},
            {"Large",  512, 2048}
    };
    ofstream cpu_out("Cpp_CPU.txt", ios::out);
    cpu_out << "Model,AvgTime_ms,PeakMemory_MB,Throughput_samples_per_sec\n";
    for (auto& cfg : configs) {
        TransformerBlock model(cfg.embed_dim, cfg.ff_hidden_dim);
        MatrixXd input = MatrixXd::Random(batch_size * seq_length, cfg.embed_dim);
        for (int i = 0; i < 10; ++i)
            model.forward(input);
        vector<double> times;
        times.reserve(runs);
        long peak_rss_kB = 0;
        for (int i = 0; i < runs; ++i) {
            auto t0 = chrono::high_resolution_clock::now();
            model.forward(input);
            auto t1 = chrono::high_resolution_clock::now();
            double ms = chrono::duration<double, milli>(t1 - t0).count();
            times.push_back(ms);
            long rss = getCurrentRSS_kB();
            if (rss > peak_rss_kB) peak_rss_kB = rss;
        }
        double avg_ms      = accumulate(times.begin(), times.end(), 0.0) / times.size();
        double peak_mem_MB = peak_rss_kB / 1024.0;
        double throughput  = batch_size / (avg_ms / 1000.0);
        cout << "=== Benchmark: " << cfg.name << " model ===\n"
             << "Average execution time: " << avg_ms      << " ms\n"
             << "Peak RSS (CPU mem):     " << peak_mem_MB << " MB\n"
             << "Throughput:             " << throughput  << " samples/sec\n\n";
        cpu_out << cfg.name << ","
                << avg_ms      << ","
                << peak_mem_MB << ","
                << throughput  << "\n";
    }
    cpu_out.close();
    return 0;
}

File: 605proj/Benchmark.py
#!/usr/bin/env python3
import os
import time
import gc
import psutil
import pandas as pd
import matplotlib.pyplot as plt
from tqdm import tqdm
import torch
import torch.nn as nn

configs = {
    'Small':  dict(embed_dim=128, num_heads=4,  ff_hidden_dim=512),
    'Medium': dict(embed_dim=256, num_heads=8,  ff_hidden_dim=1024),
    'Large':  dict(embed_dim=512, num_heads=8,  ff_hidden_dim=2048),
}
batch_size = 32
seq_length = 128
n_trials   = 50

cpp_cpu_txt = 'Cpp_CPU.txt'
cpp_gpu_txt = 'Cpp_GPU.txt'

class TransformerBlock(nn.Module):
    def __init__(self, embed_dim, num_heads, ff_hidden_dim):
        super().__init__()
        self.attn = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True)
        self.ff   = nn.Sequential(
            nn.Linear(embed_dim, ff_hidden_dim),
            nn.ReLU(),
            nn.Linear(ff_hidden_dim, embed_dim),
        )
        self.ln1  = nn.LayerNorm(embed_dim)
        self.ln2  = nn.LayerNorm(embed_dim)

    def forward(self, x):
        attn_out, _ = self.attn(x, x, x)
        x = self.ln1(x + attn_out)
        f = self.ff(x)
        return self.ln2(x + f)
def benchmark(device: torch.device):
    process = psutil.Process(os.getpid())
    results = {'Model': [], 'AvgTime(ms)': [], 'PeakMem(MB)': [], 'Throughput(samps/s)': []}

    for name, cfg in configs.items():
        model = TransformerBlock(**cfg).to(device).eval()
        x     = torch.randn(batch_size, seq_length, cfg['embed_dim'], device=device)

        with torch.no_grad():
            for _ in range(10):
                _ = model(x)

        if device.type == 'cuda':
            torch.cuda.reset_peak_memory_stats(device)
        else:
            base_rss = process.memory_info().rss
            peak_rss = base_rss

        times = []
        for _ in tqdm(range(n_trials), desc=f"{name} @ {device.type}", leave=False):
            gc.collect()
            if device.type == 'cuda':
                torch.cuda.empty_cache()

            t0 = time.perf_counter()
            with torch.no_grad():
                _ = model(x)
            t1 = time.perf_counter()
            times.append((t1 - t0) * 1000)

            if device.type == 'cpu':
                now_rss  = process.memory_info().rss
                peak_rss = max(peak_rss, now_rss)

        avg_t = sum(times) / len(times)
        thrpt = batch_size / (avg_t / 1000)

        if device.type == 'cuda':
            peak_mem = torch.cuda.max_memory_allocated(device) / (1024**2)
        else:
            peak_mem = (peak_rss - base_rss) / (1024**2)

        results['Model'].append(name)
        results['AvgTime(ms)'].append(avg_t)
        results['PeakMem(MB)'].append(peak_mem)
        results['Throughput(samps/s)'].append(thrpt)

        del model, x
        if device.type == 'cuda':
            torch.cuda.empty_cache()
        gc.collect()

    return pd.DataFrame(results)

print("Running Python CPU benchmark on cpu...")
df_py_cpu = benchmark(torch.device('cpu'))

df_py_gpu = None
if torch.cuda.is_available():
    print("\nRunning Python GPU benchmark on cuda...")
    df_py_gpu = benchmark(torch.device('cuda'))
else:
    print("\nCUDA not available: skipping Python GPU benchmark.")


df_cpp_cpu = pd.read_csv(cpp_cpu_txt) if os.path.exists(cpp_cpu_txt) else None
df_cpp_gpu = pd.read_csv(cpp_gpu_txt) if os.path.exists(cpp_gpu_txt) else None


models = list(configs.keys())
combined_time = pd.DataFrame(index=models)
combined_mem  = pd.DataFrame(index=models)
combined_thr  = pd.DataFrame(index=models)

combined_time['Python_CPU'] = df_py_cpu.set_index('Model')['AvgTime(ms)']
combined_mem ['Python_CPU'] = df_py_cpu.set_index('Model')['PeakMem(MB)']
combined_thr ['Python_CPU'] = df_py_cpu.set_index('Model')['Throughput(samps/s)']

if df_py_gpu is not None:
    combined_time['Python_GPU'] = df_py_gpu.set_index('Model')['AvgTime(ms)']
    combined_mem ['Python_GPU'] = df_py_gpu.set_index('Model')['PeakMem(MB)']
    combined_thr ['Python_GPU'] = df_py_gpu.set_index('Model')['Throughput(samps/s)']

if df_cpp_cpu is not None:
    combined_time['Cpp_CPU'] = df_cpp_cpu.set_index('Model')['AvgTime(ms)']
    combined_mem ['Cpp_CPU'] = df_cpp_cpu.set_index('Model')['PeakMem(MB)']
    combined_thr ['Cpp_CPU'] = df_cpp_cpu.set_index('Model')['Throughput(samps/s)']

if df_cpp_gpu is not None:
    combined_time['Cpp_GPU'] = df_cpp_gpu.set_index('Model')['AvgTime(ms)']
    combined_mem ['Cpp_GPU'] = df_cpp_gpu.set_index('Model')['PeakMem(MB)']
    combined_thr ['Cpp_GPU'] = df_cpp_gpu.set_index('Model')['Throughput(samps/s)']


print("\n-- Combined time table --")
print(combined_time)

print("\n-- Combined memory table --")
print(combined_mem)

print("\n-- Combined throughput table --")
print(combined_thr)


os.makedirs('charts', exist_ok=True)

fig, ax = plt.subplots(figsize=(8,5))
combined_time.plot.bar(ax=ax)
ax.set_title("Execution Time Comparison")
ax.set_ylabel("Time (ms)")
ax.set_xticklabels(models, rotation=0)
fig.tight_layout()
fig.savefig("charts/execution_time.png")

fig, ax = plt.subplots(figsize=(8,5))
combined_mem.plot.bar(ax=ax)
ax.set_title("Peak Memory Footprint (MB)")
ax.set_ylabel("Memory (MB)")
ax.set_xticklabels(models, rotation=0)
fig.tight_layout()
fig.savefig("charts/memory_comparison.png")

fig, ax = plt.subplots(figsize=(8,5))
combined_thr.plot.bar(ax=ax)
ax.set_title("Throughput Comparison")
ax.set_ylabel("Samples/sec")
ax.set_xticklabels(models, rotation=0)
fig.tight_layout()
fig.savefig("charts/throughput_comparison.png")

print("\nSaved comparison charts under ./charts/")